File: app/src/main/java/id/ac/stiki/doleno/absenin/domain/impl/FetchAllEventParticipantImpl.java
package id.ac.stiki.doleno.absenin.domain.impl;

import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;

import id.ac.stiki.doleno.absenin.data.entity.EventParticipant;
import id.ac.stiki.doleno.absenin.domain.FetchAllEventParticipant;
import id.ac.stiki.doleno.absenin.repository.EventParticipantRepository;

public class FetchAllEventParticipantImpl implements FetchAllEventParticipant {
    private final EventParticipantRepository eventParticipantRepository;

    public FetchAllEventParticipantImpl(EventParticipantRepository eventParticipantRepository) {
        this.eventParticipantRepository = eventParticipantRepository;
    }

    @Override
    public Task<QuerySnapshot> execute(int eventId) {
        return eventParticipantRepository.get(eventId)
                .addOnSuccessListener(result -> {
                    List<EventParticipant> eventParticipants = new ArrayList<>();
                    result.getDocuments().forEach(document -> eventParticipants.add(new EventParticipant(document.getData())));
                    eventParticipantRepository.create(eventParticipants);
                });
    }
}

File: app/src/main/java/id/ac/stiki/doleno/absenin/domain/impl/FetchAllEventParticipantImpl.kt
package id.ac.stiki.doleno.absenin.domain.impl

import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.QuerySnapshot
import id.ac.stiki.doleno.absenin.data.entity.EventParticipant
import id.ac.stiki.doleno.absenin.domain.FetchAllEventParticipant
import id.ac.stiki.doleno.absenin.repository.EventParticipantRepository
import java.util.concurrent.Executors
import javax.inject.Inject

class FetchAllEventParticipantImpl @Inject constructor(private val eventParticipantRepository: EventParticipantRepository) :
    FetchAllEventParticipant {
    override fun execute(eventId: Long): Task<QuerySnapshot> {
        return eventParticipantRepository.get(eventId)
            .addOnSuccessListener { result: QuerySnapshot ->
                Executors.newSingleThreadExecutor().execute {
                    val eventParticipants: List<EventParticipant> = result.documents.map {
                        EventParticipant(it.data)
                    }
                    eventParticipantRepository.create(eventParticipants)
                }
            }
    }
}